addressmanager/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+use parking_lot::Mutex;
+use rocksdb::{DBWithThreadMode, MultiThreaded};
+use stores::banned_address_store::{BannedAddressesStore, BannedAddressesStoreReader, DbBannedAddressesStore};
+
+const MAX_ADDRESSES: usize = 4096;
+const MAX_CONNECTION_FAILED_COUNT: u64 = 3;
+pub type DB = DBWithThreadMode<MultiThreaded>;
+
+pub struct AddressManager {
+    banned_address_store: DbBannedAddressesStore,
+    not_banned_address_store: not_banned_address_store_with_cache::Store,
+}
+
+#[derive(PartialEq, Eq, Hash, Copy, Clone)]
+pub struct NetAddress {
+    ip: Ipv6Addr,
+    port: u16,
+}
+
+impl NetAddress {
+    pub fn new(ip: Ipv6Addr, port: u16) -> Self {
+        Self { ip, port }
+    }
+}
+
+impl AddressManager {
+    pub fn new(db: Arc<DB>) -> Arc<Mutex<Self>> {
+        Arc::new(Mutex::new(Self {
+            banned_address_store: DbBannedAddressesStore::new(db.clone(), MAX_ADDRESSES as u64),
+            not_banned_address_store: not_banned_address_store_with_cache::new(db),
+        }))
+    }
+
+    pub fn add_address(&mut self, address: NetAddress) {
+        // TODO: Don't add non routable addresses
+
+        // We mark `connectionFailedCount` as 0 only after first success
+        self.not_banned_address_store.set(address, 1);
+    }
+
+    pub fn mark_connection_failure(&mut self, address: NetAddress) {
+        let new_count = self.not_banned_address_store.get(address) + 1;
+        if new_count > MAX_CONNECTION_FAILED_COUNT {
+            self.not_banned_address_store.remove(address);
+        } else {
+            self.not_banned_address_store.set(address, new_count);
+        }
+    }
+
+    pub fn mark_connection_success(&mut self, address: NetAddress) {
+        self.not_banned_address_store.set(address, 0);
+    }
+
+    pub fn get_all_addresses(&self) -> impl Iterator<Item = NetAddress> + '_ {
+        self.not_banned_address_store.get_all_addresses()
+    }
+
+    pub fn get_random_addresses(&self, count: usize, exceptions: HashSet<NetAddress>) -> Vec<NetAddress> {
+        self.not_banned_address_store.get_random_addresses(count, exceptions)
+    }
+
+    pub fn ban(&mut self, ip: Ipv6Addr) {
+        self.banned_address_store.set(ip, unix_time()).unwrap();
+        self.not_banned_address_store.remove_by_ip(ip);
+    }
+
+    pub fn unban(&mut self, ip: Ipv6Addr) {
+        self.banned_address_store.remove(ip).unwrap();
+    }
+
+    pub fn is_banned(&mut self, ip: Ipv6Addr) -> bool {
+        const MAX_BANNED_TIME: u64 = 24 * 60 * 60 * 1000;
+        match self.banned_address_store.get(ip).unwrap_option() {
+            Some(timestamp) => {
+                if unix_time() - timestamp > MAX_BANNED_TIME {
+                    self.unban(ip);
+                    false
+                } else {
+                    true
+                }
+            }
+            None => false,
+        }
+    }
+}
+
+fn unix_time() -> u64 {
+    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64
+}
+
```

Comment:
Add a comment explaining that here we want the full store to be in memory and we don't mind assuming single-thread access since this is not an intensive operation